### PR TITLE
SMP Playlist Helper, controller, route and tests.

### PR DESCRIPTION
**Pull request by BjornTwachtmann**

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/DsShared/Helpers/SmpPlaylistHelper.php [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+use BBC\ProgrammesPagesService\Domain\Entity\MusicSegment;
+use BBC\ProgrammesPagesService\Domain\Entity\Programme;
+use BBC\ProgrammesPagesService\Domain\Entity\ProgrammeItem;
+use BBC\ProgrammesPagesService\Domain\Entity\SegmentEvent;
+use BBC\ProgrammesPagesService\Domain\Entity\Version;
+use BBC\ProgrammesPagesService\Domain\Entity\VersionType;
+
+class SmpPlaylistHelper
+{
+    private const DEFAULT_WARNING_VPID = 'p025x55x';
```
Comment (votemike) [inline src/DsShared/Helpers/SmpPlaylistHelper.php]: What's this? Should it have a docblock explaining what it is?

File: src/DsShared/Helpers/SmpPlaylistHelper.php [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    /**
+     * @param ProgrammeItem $programmeItem
+     * @param Version|null $streamableVersion
+     * @param Version[] $segmentEvents
```
Comment (votemike) [inline src/DsShared/Helpers/SmpPlaylistHelper.php]: Is $segmentEvents an array of `Version`s?

File: src/DsShared/Helpers/SmpPlaylistHelper.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            'kind' => 'warning',
+        ];
+    }
+
+    /**
+     * @param SegmentEvent[] $segmentEvents
+     * @param ProgrammeItem $programmeItem
+     * @return array
+     */
+    private function getMarkers(array $segmentEvents, ProgrammeItem $programmeItem): array
```
Comment (votemike) [inline src/DsShared/Helpers/SmpPlaylistHelper.php]: This method isn't the easiest to read. Is it worth refactoring a bit?

File: tests/DataFixtures/PagesService/SegmentEventsFixture.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
+<?php
+declare(strict_types = 1);
+
+namespace Tests\App\DataFixtures\PagesService;
+
+use App\Builders\MusicSegmentBuilder;
+use App\Builders\SegmentBuilder;
+use App\Builders\SegmentEventBuilder;
+use BBC\ProgrammesPagesService\Data\ProgrammesDb\Entity\SegmentEvent;
```
Comment (Tasiobg) [inline tests/DataFixtures/PagesService/SegmentEventsFixture.php]: Do you need this lines?
use BBC\ProgrammesPagesService\Data\ProgrammesDb\Entity\SegmentEvent;
use BBC\ProgrammesPagesService\Domain\Entity\Segment;

File: tests/DsShared/Helpers/SmpPlaylistHelper/SmpPlaylistHelperTest.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
+<?php
+declare(strict_types = 1);
+
+namespace Tests\App\DsShared\Helpers\SmpPlaylistHelper;
+
+use App\Builders\EpisodeBuilder;
```
Comment (Tasiobg) [inline tests/DsShared/Helpers/SmpPlaylistHelper/SmpPlaylistHelperTest.php]: Do you need this lines?
use App\Builders\EpisodeBuilder;
use App\Builders\ImageBuilder;
use Tests\App\DataFixtures\PagesService\EpisodesFixtures;

File: src/Builders/ClipBuilder.php [hunk 6]
```diff
@@ -33,6 +33,7 @@ protected function __construct()
             'mediaType' => $faker->randomElement(MediaTypeEnum::validValues()),
             'segmentEventCount' => $faker->numberBetween(0, 8),
             'aggregatedGalleriesCount' => $faker->numberBetween(1, 5),
+            'isExternallyEmeddable' => $faker->boolean,
```
Comment (votemike) [inline src/Builders/ClipBuilder.php]: isExternallyEmeddable -> isExternallyEmbeddable

File: src/Builders/EpisodeBuilder.php [hunk 7]
```diff
@@ -35,6 +35,7 @@ protected function __construct()
             'aggregatedBroadcastsCount' => $faker->numberBetween(1, 5),
             'availableClipsCount' => $faker->numberBetween(1, 5),
             'aggregatedGalleriesCount' => $faker->numberBetween(1, 5),
+            'isExternallyEmeddable' => $faker->boolean,
```
Comment (votemike) [inline src/Builders/EpisodeBuilder.php]: isExternallyEmeddable -> isExternallyEmbeddable

File: src/Controller/SmpPlaylist/SmpPlaylistController.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        $allStreamableVersions = $versionsService->findAllStreamableByProgrammeItem($programmeItem);
+        $streamableVersion = empty($allStreamableVersions) ? null : $allStreamableVersions[0];
+
+        $segmentEvents = [];
+        if ($programmeItem->getSegmentEventCount()) {
+            $segmentEvents = $segmentEventsService->findByProgrammeForCanonicalVersion($programmeItem);
+        }
+        $smpPlaylistHelper = $helperFactory->getSmpPlaylistHelper();
+        $array = $smpPlaylistHelper->getPlaylist($programmeItem, $streamableVersion, $segmentEvents, $allStreamableVersions);
```
Comment (votemike) [inline src/Controller/SmpPlaylist/SmpPlaylistController.php]: `$array` is a bad variable name.

File: src/DsShared/Helpers/SmpPlaylistHelper.php [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                foreach ($allStreamableVersions as $version) {
+                    $feed['allAvailableVersions'][] = $this->getVersionFeed($programmeItem, $version);
+                }
+            }
+        }
+
+        return $feed;
+    }
+
+    private function getVersionFeed(ProgrammeItem $programmeItem, Version $version)
```
Comment (votemike) [inline src/DsShared/Helpers/SmpPlaylistHelper.php]: Return typehints?

**Comment by BjornTwachtmann** [comment]
@votemike , let me know if you're happy with this now

**Pull request merged**